cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -416,15 +439,29 @@ public void upgradeApplication(HttpRequest request, HttpResponder responder,
   @Path("/upgrade")
   @AuditPolicy(AuditDetail.REQUEST_BODY)
   public void upgradeApplications(FullHttpRequest request, HttpResponder responder,
-                                  @PathParam("namespace-id") String namespace) throws Exception {
+                                  @PathParam("namespace-id") String namespaceId,
+                                  @Nullable @QueryParam("artifactScope") Set<String> artifactScopes,
+                                  @QueryParam("allowSnapshot") boolean allowSnapshot) throws Exception {
     // TODO: (CDAP-16910) Improve batch API performance as each application upgrade is an event independent of each
     //  other.
-    List<ApplicationId> appIds = decodeAndValidateBatchApplication(validateNamespace(namespace), request);
+
+    Set<ArtifactScope> artifactScopesToConsider = new HashSet<>();
```

Comment:
Refactor this block to a private method. You can modify the `validateScope` to `validateScopes` for that purpose.